File: Models/Product.cs
namespace POS.Models;

public class Product
{
    public Guid productId { get; }
    public string sku { get; }
    public string name { get; }
    public float price { get; }
    public string? description { get; }
    public Category category { get; }

    public Product(string sku, string name, float price, string? description)
    {
        // Initializes product object
    }
}

public class OptionItem
{

    public Guid optionItemId { get; }
    public string name { get; }
    public float price { get; }

    public OptionItem(string name, float price)
    {
        // Initializes OptionItem object
    }
}

public class OptionGroup
{
    public string optionGroupId { get; }
    public List<OptionItem> optionItems;
    public OptionGroup(List<OptionItem> optionItems)
    {
        // Initializes OptionGroup object
    }
}

File: product-service/src/main/java/solutions/skydev/pos/product_service/model/entity/Product.java
package solutions.skydev.pos.product_service.model.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

import java.math.BigDecimal;
import java.util.HashSet;
import java.util.Set;

@Entity
@Getter
@Setter
@NoArgsConstructor
@SQLDelete(sql = "UPDATE product SET deleted = true WHERE id=?")
@SQLRestriction("deleted=false")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(precision =  7, scale = 2)
    private BigDecimal price;

    private String name;

    @Column(length=500)
    private String description;
    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "product_category",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private Set<Category> categories = new HashSet<>();

    @Column(columnDefinition = "BOOLEAN DEFAULT false")
    private Boolean deleted = Boolean.FALSE;
}
